text/1008-tls-and-alpn.rst

```diff
@@ -0,0 +1,458 @@
+corresponding EdgeDB configs mentioned in previous chapter, together
+with the other minor tunings for ``ssl.SSLContext``.
+
+``--tls-certfile``, ``--tls-keyfile`` are directly the parameters of
+``ssl.SSLContext.load_cert_chain()``, while the EdgeDB server would
+accept a password for the private key as an environment variable
+``EDGEDB_TLS_PRIVATE_KEY_PASSWORD``. However, the ``password`` argument
+of ``load_cert_chain()`` must always be set to a Python function to
+avoid triggering OpenSSL to prompt for password. If the env var is not
+set, simply return ``b""`` in the function - it will not be invoked if
+the private key is not protected by a password.
+
+When ``--tls-certfile`` is not present and the server is in ``devmode``
+or ``testmode``, the server will use the CLI to generate a self-signed
+certificate and use it to run the TLS server for development and
+testing. The generated certificate (with its private key) should be
+written to a ``credentials.json`` file just like the regular CLI
+behavior, together with the connection parameters. For example, the
+server would run this "hidden" CLI command on start::
+
+    edgedb _dev --port 5656 --user edgedb \
+        --password login-password-in-clear-text --database edgedb
+
+And it overwrites ``~/.edgedb/credentials/local-dev.json`` with::
+
+    {
+        "port": 5656,
+        "user": "edgedb",
+        "password": "login-password-in-clear-text",
+        "database": "edgedb",
+        "tls_certdata": "-----BEGIN RSA PRIVATE KEY-----\nMIIEowIBAA..."
+    }
+
+And echo back the generated certificate in standard output so that the
+server could simply read and use. So that the EdgeDB developer could
+always use ``edgedb -Ilocal-dev`` to access the dev server. The test
+suite could also take advantage from this mimic of real-life EdgeDB
+scenario to cover some real cases.
+
+Another server-side topic that was discussed in this RFC is the UNIX
+domain socket. It is proposed that the non-admin UNIX socket support
+should be removed, while the admin UNIX socket remains in clear-text
+binary protocol.
+
+
+Client Certificate
+==================
+
+Supporting client certificate authentication is a nice-to-have feature
+in this RFC, as implementing a proper client certificate authentication
+system can be complicated - if we also issue the client certificates,
+we'd probably reconsider the CA idea below. In this section, we're only
+discussing the feasibility.
+
+First of all, we'd want to add a new Auth method ``Certificate`` beyond
+the other two methods ``Trust`` and ``SCRAM``. The ``Certificate``
+``Auth`` entry tells the EdgeDB server which users are allowed to
+authenticate themselves using a client certificate.
+
+Then we would need a new config type to store the trusted client
+certificate fingerprints. The idea is to let the CLI generate the client
+certificates using a local CA, and the CLI tells the server to trust the
+generated certificates.
```

Comment:
We don't need storing client certificate fingerprints.

The idea of client certificates is:
1. We have a CA, certificate of which is know to the EdgeDB server
2. CA emits a client certificate, having some identifier of the user name, say common-name of `username.yourdb.example.org`
3. When client authenticates it sends the certificate from (2) to the EdgeDB server, server verifies it via certificate described in (1) [this involves client-certificate private key to encode some challenge-responses but I'm omitting the details now].

So on each authorization, client certificate is checked against certificate chain and no individual user info is stored between sesesions. (Well, there is TSL session cache, but that's a different layer of abstraction).